Added fcntl, fixed errno in socketpair and pipe2, added linger=0 for pipes

nikarh opened the pull request:
- Fixed error handling in `socketpair`. I forgot that `close` calls reset `errno`, so now we're caching it in a var and restoring after `close`
- `pipe2` now also sets linger to zero for underlying sockets. Why? Due to how TCP works after a TCP socket is closed, the closing part of the pair remains in `TIME_WAIT` for 60 seconds, meaning if too many pipes/socketpairs are created too fast, they deplete kernel memory for the buffers.
- `pipe2` always creates socketpair now and doesn't fallback to `pipe` for non-block. This is done for consistency - so that the consumer can use `fcntl` to switch non-block state and it works regardless of how pipe was created (but only for `pipe2` call)
- Implemented `fcntl` for `F_DUPFD`, `F_GETFD` (always 0 since there's no process API and no cloexec), `F_GETFL` and `F_SETFL` & `O_NONBLOCK` for socket fds.

Regarding `F_DUPFD` - I wanted to respect the argument, but taking into account that `MAX_OPEN_FILES` is defined as 256, and `F_DUPFD` arg is usually used to set it above 256, do we really need that? Or can we just ignore the arg and yolo delegate to `dup` call?

d3m3vilurr commented on newlib/libc/sys/vita/fcntl.c:
```diff
@@ -0,0 +1,120 @@
+
+			return 0;
+		}
+	}
```
do we need

```
else if (cmd == F_GETFL || cmd == F_SETFL) {
  errno = EINVAL;
  reutrn -1;
}
```
??

nikarh replied:
🤷‍♂️ Do you mean for file descriptors other than sockets? The POSIX fcntl `F_GETFL` and `F_SETFL` usually support regular files (e.g. `O_APPEND`), but I just didn't implement that, so they all will fall to `ENOTSUP`. 

But I can change that to `EINVAL`.

nikarh replied:
Though it still seems to me that `ENOTSUP` is more appropriate than `EINVAL`, since this is caused by the `fcntl` implementation not supporting some argument combinations, and not the argument being invalid

d3m3vilurr replied:
oops i forgot this thread.

i just followed [spec document](https://pubs.opengroup.org/onlinepubs/9699919799/functions/fcntl.html).
meaning of `ENOTSUP` is good for me. just spec doesn't defined error of `ENOTSUP`. so i thought `EBADF` and `EINVAL` for this case.

anyway i already merged this PR. we can think it with more time. :)

d3m3vilurr reviewed (changes requested):
can you split a commit to 3~4?
- minfd of dup descriptor
- patch part of pipe2
- patch part of socketpair
- implement fcntl

nikarh commented:
> can you split a commit to 3~4?
> 
>     * minfd of dup descriptor
> 
>     * patch part of pipe2
> 
>     * patch part of socketpair
> 
>     * implement fcntl

Sure, on it

nikarh committed:
Fixed error handling in socketpair call

`close` calls reset `errno` if they succeed, so changes in this PR cache
`errno` from the original error and restore it after close

nikarh committed:
Fixed pipe2 call

- Fixed error handling, same as in `socketpair`, `close` resets `errno`.
- Setting linger to 0. Otherwise when both sides of the pair are closed,
  one side remains in `TIME_WAIT` state for 60 seconds, and if too many
  pipes are opened too fast, this depletes kernel memory for the
  buffers.
- `pipe2` now always uses the socketpair implementation, even for blocking
  pipes. This is done for consistency, so that a pipe can be created,
  and later set as nonblock

nikarh committed:
Added a minfd argument to __vita_duplicate_descriptor

This change is for `fcntl` implementation.
fcntl with `F_DUPFD` has an argument which defines the minimum
file descriptor to be used for the dup.

nikarh committed:
Implemented fcntl

This is a barebones implementation that is made mainly for
getting and setting O_NONBLOCK for sockets.

nikarh commented:
Separated into commits. Is it OK to leave them all in this PR or should I separate it into 4 PRs?

d3m3vilurr commented:
> Separated into commits. Is it OK to leave them all in this PR or should I separate it into 4 PRs?

yep. it's ok.

> MAX_OPEN_FILES is defined as 256, and F_DUPFD arg is usually used to set it above 256, do we really need that? Or can we just ignore the arg and yolo delegate to dup call?

AFAIK, 256 is same as max fd limitation of vita system.

anyway, [in the spec,](https://pubs.opengroup.org/onlinepubs/9699919799/functions/fcntl.html), `F_DUPFD` only defined;
`it shall be the lowest numbered available file descriptor greater than or equal to the third argument`

so imo, when we call
```
fcntl(fd, F_DUPFD, 10);
```
return fd should be between 10 ~ 255. so i think, your implementation is quite normal.
